plugins/source.go

```diff
@@ -29,8 +26,22 @@ type SourcePlugin struct {
 	newExecutionClient SourceNewExecutionClientFunc
 	// Tables is all tables supported by this source plugin
 	tables schema.Tables
+	// status sync metrics
+	metrics SourceMetrics
+	// Logger to call, this logger is passed to the serve.Serve Client, if not define Serve will create one instead.
+	logger zerolog.Logger
+	// resourceSem is a semaphore that limits the number of concurrent resources being fetched
+	resourceSem *semaphore.Weighted
+	// tableSem is a semaphore that limits the number of concurrent tables being fetched
+	tableSem *semaphore.Weighted
+	// maxDepth is the max depth of tables
+	maxDepth uint64
 }
 
+const (
+	maxAllowedDepth = 4
```

Comment:
I would start with enforcing to make sure users/authors are following best practices of flattening/not-nesting tables. If someone will hit a blocker we can change that.

The issue is that DFS scheduler might become inefficient with big `h` which we want to avoid. and in general using nested tables if possible to use APIs directly makes plugin inefficient unrelated to any scheduler.